### ISSUE TITLE
Incorrect type inference (regression when updating from 1.1.311 -> 1.333)

### ISSUE THREAD

**Issue by mcobzarenco** [microsoft/pyright#6229]
When updating from 1.1.311 -> 1.333 there a number of new errors which I reduced to this minimal reproduction where `x` is inferred to be `tuple[float, float] | int` instead of `tuple[float, float]`, e.g.

```python
def foo(x: float | tuple[float, float]) -> float:
    if isinstance(x, float):
        x = (x, x)
    reveal_type(x)
    return x[0]
```

Running `pyright foo.py` in the command line:

```
foo.py
  foo.py:4:17 - information: Type of "x" is "tuple[float, float] | int"
  foo.py:5:12 - error: "__getitem__" method not defined on type "int" (reportGeneralTypeIssues)
1 error, 0 warnings, 1 information
```

**Comment by erictraut** [comment]
The new behavior (in 1.333) is correct. This change was introduced as part of a bug fix.

According to PEP 484, the type annotation `float` implies `float | int` even though `float` and `int` have no formal overlap (i.e. `int` is not a subclass of `float`). When you use an `isinstance` on a value that is annotated as `float`, you are checking specifically for `float`, not `int`.

That means `foo(0)` is a legal call to your function (i.e. the call will not generate a type error), but the function `foo` will crash if you attempt to make this call. Newer versions of pyright correctly detect this programming error whereas older versions did not.

**Comment by mcobzarenco** [comment]
I see, today I learnt.. thank you for the explanation.
